### PR TITLE
sway-beta: init at 1.0-beta1

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change
~~Based on the meson and systemd update from https://github.com/NixOS/nixpkgs/pull/46020, reccomend merging that first.~~ Not anymore, due to f4615bef063265001f47ee7e8e351aa1708f773a

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by danieldk** [comment]
Note that the maintainer asks distributions not to package the betas (as with previous alphas):

*Package maintainers: please do not package this release.*

Source: https://github.com/swaywm/sway/releases

Can't wait until Sway hits 1.0 either though ;).

**Comment by Synthetica9** [comment]
I preemptively asked for the author's position on that in the freenode/#sway irc channel:

> <Synthetica> ddevault: Extremely late reaction, but you said "we won't be expecting package maintainers to package anything until 1.0 proper". Does that mean you will actively discourage packaging it like in alpha, or would a seperate sway-beta package be fine?
> <ddevault> Synthetica: I wouldn't mind a beta package if it's clear that it's not considered stable

**Comment by danieldk** [comment]
Nice!

File: pkgs/development/libraries/wlroots/default.nix [hunk 1]
```diff
@@ -1,20 +1,30 @@
...
 , wayland, libGL, wayland-protocols, libinput, libxkbcommon, pixman
 , xcbutilwm, libX11, libcap, xcbutilimage, xcbutilerrors, mesa_noglu
+, x11Support ? true, xorg ? null, xcb ? null
 }:
 
 let pname = "wlroots";
-    version = "unstable-2018-03-16";
-in stdenv.mkDerivation rec {
+    version = "0.1";
+in stdenv.mkDerivation {
```
Comment (dywedir) [inline pkgs/development/libraries/wlroots/default.nix]:
```suggestion
in stdenv.mkDerivation rec {
```

**Comment by Synthetica9** [comment]
@dywedir Why do you suggest adding the `rec`?

File: pkgs/development/libraries/wlroots/default.nix (hunk 1 above)
Comment (Synthetica9) [reply pkgs/development/libraries/wlroots/default.nix]: Primeos just updated the whole file, using that patch.

**Comment by xeji** [comment]
The dependency `wlroots` is marked as broken (apparently at upstream request).

**Comment by colemickens** [comment]
@xeji  this was the wlroots change (I can't see/find the PR?) https://github.com/NixOS/nixpkgs/commit/f4615bef063265001f47ee7e8e351aa1708f773a

I'm curious if upstream is requesting that `wlroots` stays marked as broken. It seems like upstream made the decision arbitrarily, but wlroots 0.1 is at least a real release.

**Comment by xeji** [comment]
wlroots 0.1 is a tagged release, and the upstream author https://github.com/NixOS/nixpkgs/issues/38344#issuecomment-432032834 doesn't mind us trying it. So I think we can remove the `broken=true`.

**Comment by Synthetica9** [comment]
@xeji #48869

**Comment by xeji** [comment]
@GrahamcOfBorg build sway-beta

**Comment by xeji** [comment]
@GrahamcOfBorg build sway

File: pkgs/applications/window-managers/sway/default.nix [hunk 2]
```diff
@@ -1,40 +1,62 @@
...
+    ] ++ stdenv.lib.optional buildDocs [ asciidoc libxslt docbook_xsl
+    ] ++ stdenv.lib.optional cmakeBuild [ cmake
+    ] ++ stdenv.lib.optional mesonBuild [ meson ninja
+    ];
+
+
+    buildInputs = [
+      wayland wlc libxkbcommon pcre json_c dbus
+      pango cairo libinput libcap pam gdk_pixbuf libpthreadstubs
+      libXdmcp wlroots wayland-protocols git
```
Comment (xeji) [inline pkgs/applications/window-managers/sway/default.nix]: I didn't look at the source, but having `git` in `buildInputs` looks unusual for this kind of package. Why is this required?

Comment (Synthetica9) [reply pkgs/applications/window-managers/sway/default.nix]: It fails without it (last few lines of log):

```
Library pam found: YES
Library crypt found: YES
Native dependency libsystemd found: YES 239
Dependency libelogind found: NO
Library m found: YES
Library rt found: YES
Program git found: NO
Native dependency xcb found: YES 1.13.1
Program scdoc found: NO

meson.build:130:2: ERROR:  Problem encountered: git is required to make the version string

A full log can be found at /build/source/build/meson-logs/meson-log.txt
builder for '/nix/store/phhphsml185jdmz7f13yh3ic7xpcgb4k-sway-beta-1.0-beta.1.drv' failed with exit code 1
error: build of '/nix/store/phhphsml185jdmz7f13yh3ic7xpcgb4k-sway-beta-1.0-beta.1.drv' failed
```

Comment (xeji) [reply pkgs/applications/window-managers/sway/default.nix]: Is it actually used at runtime, or can we put it in `nativeBuildInputs`?

Comment (Synthetica9) [reply pkgs/applications/window-managers/sway/default.nix]: Oh, yeah, I think `nativeBuildInputs` would be fine as well.

**Review by xeji — APPROVED** [review]

**Comment by colemickens** [comment]
Git isn't really required. You can set 'sway-version' in meson flags. I'm
not sure how it would work just including git since it won't have the git
directory to determine the version tag/commit.

**Comment by primeos** [comment]
Hey *, sorry I wasn't aware of this PR @colemickens thanks for CCing me :)

@Synthetica9 thanks for asking Drew, you can remove `meta.broken = true` then.

**Comment by Synthetica9** [comment]
> Git isn't really required. You can set 'sway-version' in meson flags. I'm not sure how it would work just including git since it won't have the git directory to determine the version tag/commit.

See 7307d80 

> @Synthetica9 thanks for asking Drew, you can remove meta.broken = true then.

See https://github.com/NixOS/nixpkgs/pull/48869

**Review by primeos — CHANGES_REQUESTED** [review]
Would you mind creating a new file `beta.nix` (or something similar) for Sway 1.0? In general it is a good idea to avoid redundancy but in this case we don't need to maintain 0.15.x for long and the build system changed. IMO it would therefore be cleaner to use a separate file (even though there will be some redundancy).

Instead of Git (not required) you can use the following: `mesonFlags = [ "-Dsway-version=${version}" ];`. And the build workflow for the man pages changed as well (they now require `scdoc`). You can have a look at https://github.com/primeos/nixpkgs-primeos/blob/master/default.nix#L75 if you want (it's probably not perfect, as I didn't check if all dependencies are still required (etc.), but it works).

**Comment by Synthetica9** [comment]
> Would you mind creating a new file beta.nix (or something similar) for Sway 1.0? In general it is a good idea to avoid redundancy but in this case we don't need to maintain 0.15.x for long and the build system changed. IMO it would therefore be cleaner to use a separate file (even though there will be some redundancy).

Sure thing, working on it.

> Instead of Git (not required) you can use the following: mesonFlags = [ "-Dsway-version=${version}" ];. And the build workflow for the man pages changed as well (they now require scdoc). You can have a look at https://github.com/primeos/nixpkgs-primeos/blob/master/default.nix#L75 if you want (it's probably not perfect, as I didn't check if all dependencies are still required (etc.), but it works).

Did that in 7307d80

File: pkgs/applications/window-managers/sway/default.nix [hunk 3]
```diff
@@ -1,40 +1,63 @@
...
+    ];
+
+    enableParallelBuilding = true;
+
+    cmakeFlags = "-DVERSION=${version} -DLD_LIBRARY_PATH=/run/opengl-driver/lib:/run/opengl-driver-32/lib:";
+    mesonFlags = "-Dsway-version=${version}";
+
+    meta = with stdenv.lib; {
+      description = "i3-compatible window manager for Wayland";
+      homepage    = http://swaywm.org;
```
Comment (primeos) [inline pkgs/applications/window-managers/sway/default.nix]: You can also replace this link with https://swaywm.org/ if you want ;)

File: pkgs/applications/window-managers/sway/beta.nix [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  src = fetchFromGitHub {
+    owner = "swaywm";
+    repo = "sway";
+    rev = version;
+    sha256 = "0h9kgrg9mh2acks63z72bw3lwff32pf2nb4i7i5xhd9i6l4gfnqa";
+  };
+
+  nativeBuildInputs = [
+    pkgconfig meson ninja
+  ] ++ stdenv.lib.optional buildDocs [ asciidoc libxslt docbook_xsl ];
```
Comment (primeos) [inline pkgs/applications/window-managers/sway/beta.nix]: `[ asciidoc libxslt docbook_xsl ]` can be replaced with `[ scdoc ]`

File: pkgs/applications/window-managers/sway/beta.nix [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+, wayland, wlc, libxkbcommon, pcre, json_c, dbus
+, pango, cairo, libinput, libcap, pam, gdk_pixbuf, libpthreadstubs
+, libXdmcp, wlroots, wayland-protocols
+, buildDocs ? true
+}:
+
+stdenv.mkDerivation rec {
+  name = "${pname}-${version}";
+  pname = "sway-beta";
+  version = "1.0-beta.1";
```
Comment (primeos) [inline pkgs/applications/window-managers/sway/beta.nix]: I would probably drop "-beta" from pname, because it's already in version.

File: pkgs/applications/window-managers/sway/beta.nix [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+  nativeBuildInputs = [
+    pkgconfig meson ninja
+  ] ++ stdenv.lib.optional buildDocs [ asciidoc libxslt docbook_xsl ];
+
+  buildInputs = [
+    wayland wlc libxkbcommon pcre json_c dbus
+    pango cairo libinput libcap pam gdk_pixbuf libpthreadstubs
+    libXdmcp wlroots wayland-protocols
+  ];
```
Comment (primeos) [inline pkgs/applications/window-managers/sway/beta.nix]: You can drop `wlc`. I'm not sure if `libXdmcp` and `libpthreadstubs` are still required (it builds fine without them but I didn't verify if they're really not required - they aren't listed [here](https://github.com/swaywm/sway#compiling-from-source) as well).

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -17190,7 +17190,9 @@ with pkgs;
   wlroots = callPackage ../development/libraries/wlroots { };
   rootston = wlroots.bin;
   orbment = callPackage ../applications/window-managers/orbment { };
-  sway = callPackage ../applications/window-managers/sway { };
+  swayPkgs = callPackage ../applications/window-managers/sway { };
+
+  inherit (swayPkgs) sway sway-beta;
```
Comment (primeos) [inline pkgs/top-level/all-packages.nix]: Must be updated (sway + sway-beta).

File: pkgs/applications/window-managers/sway/beta.nix [hunk 8]
```diff
@@ -0,0 +1,43 @@
+{ stdenv, fetchFromGitHub
+, meson, ninja
+, pkgconfig, asciidoc, libxslt, docbook_xsl
+, wayland, wlc, libxkbcommon, pcre, json_c, dbus
+, pango, cairo, libinput, libcap, pam, gdk_pixbuf, libpthreadstubs
+, libXdmcp, wlroots, wayland-protocols
+, buildDocs ? true
+}:
```
Comment (primeos) [inline pkgs/applications/window-managers/sway/beta.nix]: Don't forget to remove unused attributes ;)

**Review by primeos — CHANGES_REQUESTED** [review]
Just as some sort of checklist ;) (you're probably still working on it)

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (xeji) [inline pkgs/top-level/all-packages.nix]: This doesn't fit any more -> ofborg eval error

Comment (xeji) [reply pkgs/top-level/all-packages.nix]: nevermind, @primeos was a few seconds faster :smile:

Comment (Synthetica9) [reply pkgs/top-level/all-packages.nix]: Oh, whoops, forgot to check that in :sweat_smile:

**Review by primeos — APPROVED** [review]
Looks good so far, thanks :)

**Comment by primeos** [comment]
I'll test it quickly, mind if I squash and merge to have a clean history?

**Comment by primeos** [comment]
Ok, it worked fine :)
